### PR TITLE
DOMA-1309 added virtual inn field with validation dependent on organization country code

### PR DESCRIPTION
**Pull request by akarjakin**

### CODE REVIEW

File: apps/condo/domains/common/hooks/useValidations.ts [hunk 1]
```diff
@@ -33,6 +35,8 @@ export const useValidations: UseValidations = (settings = {}) => {
     const FieldIsTooShortMessage = intl.formatMessage({ id: 'ValueIsTooShort' })
     const FieldIsTooLongMessage = intl.formatMessage({ id: 'ValueIsTooLong' })
     const NumberIsNotValidMessage = intl.formatMessage({ id: 'NumberIsNotValid' })
+    const InnValueIsInvalid = intl.formatMessage({ id: 'pages.organizations.inn.InvalidValue' })
```
Comment (MrFoxPro) [inline apps/condo/domains/common/hooks/useValidations.ts:38]: Add Message Suffix

File: apps/condo/domains/common/utils/codegeneration/generate.gql.js [hunk 2]
```diff
@@ -66,8 +66,10 @@ function genDeleteGQL (key, fields) {
 
 function generateGqlQueries (key, fields) {
     const [singularName, pluralName] = getModelForms(key)
+
     if (!fields.startsWith('{') || !fields.endsWith('}'))
         throw new Error('wrong list fields format. Try "{ name1 name2 }"')
+
```
Comment (MrFoxPro) [inline apps/condo/domains/common/utils/codegeneration/generate.gql.js:72]: Why new lines here?

File: apps/condo/domains/common/utils/validation.utils.js [hunk 3]
```diff
@@ -1,112 +1,211 @@
...
     if (typeof resolvedData === 'undefined') throw new Error('unexpected undefined resolvedData arg')
     if (typeof existingItem === 'undefined') existingItem = {}
+
     let hasAllFields = true
+
     for (let field of databaseRequired) {
-        let value = existingItem[field]
-        let isValueNullOrUndefined = value === null || typeof value === 'undefined'
+        const value = existingItem[field]
+        const isValueNullOrUndefined = value === null || typeof value === 'undefined'
```
Comment (MrFoxPro) [inline apps/condo/domains/common/utils/validation.utils.js:20]: can we replace it with lodash utils?

File: apps/condo/domains/common/hooks/useValidations.ts (hunk 1 above)
Comment (akarjakin) [reply apps/condo/domains/common/hooks/useValidations.ts:38]: done

File: apps/condo/domains/common/utils/validation.utils.js [hunk 4]
```diff
@@ -1,112 +1,215 @@
...
             addFieldValidationError(`${REQUIRED_NO_VALUE_ERROR}${field}] Value is not set`)
             hasAllFields = false
         }
     }
+
     return hasAllFields
 }
+
+const DV_FIELD_NAME = 'dv'
+const SENDER_FIELD_NAME = 'sender'
```
Comment (dkoviazin) [inline apps/condo/domains/common/utils/validation.utils.js:36]: Better to move to contants

File: apps/condo/domains/common/utils/validation.utils.js [hunk 5]
```diff
@@ -1,112 +1,215 @@
...
+const VALID_JSON_FIELD_CONSTRAINTS_FORMAT_REGEXP = /^[a-zA-Z0-9!#$%()*+-;=,:[\]/.?@^_`{|}~]{5,42}$/
```
Comment (dkoviazin) [inline apps/condo/domains/common/utils/validation.utils.js:37]: Do we need this validation?

File: apps/condo/domains/common/utils/validation.utils.js [hunk 6]
```diff
@@ -1,112 +1,215 @@
...
             }
         }
+
         return !errors
     } else {
         return addFieldValidationError(`${JSON_UNKNOWN_VERSION_ERROR}${fieldPath}] Unknown \`dv\` attr inside JSON Object`)
     }
 }
 
+const DEFAULT_COUNTRY = RUSSIA_COUNTRY
```
Comment (dkoviazin) [inline apps/condo/domains/common/utils/validation.utils.js:155]: Looks like constants

File: apps/condo/domains/common/utils/validation.utils.js [hunk 7]
```diff
@@ -1,112 +1,215 @@
...
+const RU_INN_DIGITS = [3, 7, 2, 4, 10, 3, 5, 9, 4, 6, 8, 0]
+const RU_INN_REGEXP = /^\d{10}$|^\d{12}$/
+const RU_ORGANIZATION_INN_REGEXP = /^\d{10}$/
+const VALID_RU_INN_10 = '1654019570'
```
Comment (dkoviazin) [inline apps/condo/domains/common/utils/validation.utils.js:159]: Better to move to test utils

File: apps/condo/domains/common/utils/codegeneration/generate.gql.js (hunk 2 above)
Comment (pahaz) [reply apps/condo/domains/common/utils/codegeneration/generate.gql.js:72]: looks like a “guards code block”

File: apps/condo/domains/common/utils/validation.utils.js [hunk 8]
```diff
@@ -1,112 +1,215 @@
...
+const VALID_RU_INN_12 = '500110474504'
+const INVALID_RU_INN_10 = '01234556789'
+const INVALID_RU_INN_12 = '0123455678901'
+const SOME_RANDOM_LETTERS = 'ABCDEFGHIJ'
```
Comment (pahaz) [inline apps/condo/domains/common/utils/validation.utils.js:163]: Test date, move to tests

File: apps/condo/domains/common/utils/validation.utils.js [hunk 9]
```diff
@@ -1,112 +1,215 @@
...
+}
+
+const getIsInnValid = (country = DEFAULT_COUNTRY) => (innValue = null) => isInnValid(innValue, country)
+
 module.exports = {
     hasDbFields,
     hasOneOfFields,
     hasDvAndSenderFields,
     hasValidJsonStructure,
+    validateInnRU,
```
Comment (pahaz) [inline apps/condo/domains/common/utils/validation.utils.js:207]: It’s better to extract it to inn.utils.ja

File: apps/condo/domains/organization/utils/testSchema/index.js [hunk 10]
```diff
@@ -24,21 +25,35 @@ const OrganizationLink = generateGQLTestUtils(OrganizationLinkGQL)
...
+ * @param extraAttrs
+ * @param extraMeta {inn?: string, country?: string}
+ * @returns {Promise<({data: *, errors: *}|*|{country: (*|string), dv: number, sender: {dv: number, fingerprint: *}, meta: {zipCode: *, number: *, country: (*|string), dv: number, city, street: *, inn: (*), kpp: *}, name: *, description: *})[]>}
+ */
+async function createTestOrganization (client, extraAttrs = {}, extraMeta = {}) {
     if (!client) throw new Error ('no client')
+
     const sender = { dv: 1, fingerprint: faker.random.alphaNumeric (8) }
-    const country = DEFAULT_ENGLISH_COUNTRY
+    const country = get(extraMeta, 'country') || DEFAULT_ENGLISH_COUNTRY
```
Comment (pahaz) [inline apps/condo/domains/organization/utils/testSchema/index.js:45]: It’s redundant. You can override. Look at

File: apps/condo/domains/organization/utils/testSchema/index.js [hunk 11]
```diff
@@ -24,21 +25,35 @@ const OrganizationLink = generateGQLTestUtils(OrganizationLinkGQL)
...
+}
+
+/**
+ * Creates test organization prefilled with faker data, hydrated with extraAttrs and extraMeta, if provided
+ * @param client
+ * @param extraAttrs
+ * @param extraMeta {inn?: string, country?: string}
+ * @returns {Promise<({data: *, errors: *}|*|{country: (*|string), dv: number, sender: {dv: number, fingerprint: *}, meta: {zipCode: *, number: *, country: (*|string), dv: number, city, street: *, inn: (*), kpp: *}, name: *, description: *})[]>}
+ */
+async function createTestOrganization (client, extraAttrs = {}, extraMeta = {}) {
```
Comment (pahaz) [inline apps/condo/domains/organization/utils/testSchema/index.js:41]: No it’s a bad idea to change the common signature here. If you need to change meta, you need to use the second argument

Comment (pahaz) [reply apps/condo/domains/organization/utils/testSchema/index.js:41]: It’s a common style. Don’t change it.

File: apps/condo/domains/organization/utils/testSchema/index.js [hunk 12]
```diff
@@ -280,6 +296,7 @@ module.exports = {
     OrganizationLink, createTestOrganizationLink, updateTestOrganizationLink,
     makeEmployeeUserClientWithAbilities,
     TokenSet, createTestTokenSet, updateTestTokenSet,
+    readOrganization,
```
Comment (pahaz) [inline apps/condo/domains/organization/utils/testSchema/index.js:299]: Use get All

File: apps/condo/domains/common/utils/validation.utils.js (hunk 5 above)
Comment (akarjakin) [reply apps/condo/domains/common/utils/validation.utils.js:37]: added TODO

File: apps/condo/domains/common/utils/validation.utils.js (hunk 7 above)
Comment (akarjakin) [reply apps/condo/domains/common/utils/validation.utils.js:159]: done

File: apps/condo/domains/common/utils/validation.utils.spec.js [hunk 13]
```diff
@@ -1,4 +1,13 @@
-const { hasValidJsonStructure } = require('./validation.utils')
+const {
+    hasValidJsonStructure,
+    isInnValid,
+    VALID_RU_INN_10,
```
[the remark this answers is not in the document]
Comment (akarjakin) [reply apps/condo/domains/common/utils/validation.utils.spec.js:4]: done

File: apps/condo/domains/organization/schema/Organization.js [hunk 14]
```diff
@@ -38,6 +39,27 @@ const Organization = new GQLListSchema('Organization', {
             isRequired: true,
             kmigratorOptions: { null: false },
         },
+        // The reason for this field is to avoid adding check for resident user into global Organization read access.
+        // This field have specific use case for mobile client.
+        inn: {
```
[the remark this answers is not in the document]
Comment (akarjakin) [reply apps/condo/domains/organization/schema/Organization.js:44]: fixed

File: apps/condo/domains/organization/schema/Organization.test.js [hunk 15]
```diff
@@ -1,16 +1,39 @@
...
+
+const { RUSSIA_COUNTRY } = require('@condo/domains/common/constants/countries')
+const {
+    catchErrorFrom,
+    expectToThrowAuthenticationErrorToObjects,
+    expectToThrowAccessDeniedErrorToObj,
+    expectToThrowAuthenticationErrorToObj,
+} = require('@condo/domains/common/utils/testSchema')
+const {
+    VALID_RU_INN_10,
```
[the remark this answers is not in the document]
Comment (akarjakin) [reply apps/condo/domains/organization/schema/Organization.test.js:14]: fixed

File: apps/condo/domains/organization/utils/testSchema/index.js (hunk 10 above)
Comment (akarjakin) [reply apps/condo/domains/organization/utils/testSchema/index.js:45]: removed

File: apps/condo/domains/organization/utils/testSchema/index.js (hunk 12 above)
Comment (akarjakin) [reply apps/condo/domains/organization/utils/testSchema/index.js:299]: done

File: apps/condo/domains/organization/schema/Organization.js [hunk 16]
```diff
@@ -38,6 +39,27 @@ const Organization = new GQLListSchema('Organization', {
...
+            schemaDoc: 'Organization inn',
```
[the remark this answers is not in the document]
Comment (akarjakin) [reply apps/condo/domains/organization/schema/Organization.js:45]: done

File: apps/condo/domains/organization/gql.js [hunk 17]
```diff
@@ -1,5 +1,5 @@
 /**
- * Generated by `createschema organization.Organization 'country:Select:ru,en; name:Text; description?:Text; avatar?:File; meta:Json; employees:Relationship:OrganizationEmployee:CASCADE; statusTransitions:Json; defaultEmployeeRoleStatusTransitions:Json' --force`
+ * Generated by `createschema organization.Organization 'country:Select:ru,en; name:Text; description?:Text; avatar?:File; meta:Json; inn?:String; employees:Relationship:OrganizationEmployee:CASCADE; statusTransitions:Json; defaultEmployeeRoleStatusTransitions:Json' --force`
```
Comment (pahaz) [inline apps/condo/domains/organization/gql.js:2]:
```suggestion
 * Generated by `createschema organization.Organization 'country:Select:ru,en; name:Text; description?:Text; avatar?:File; meta:Json; tin?:String; employees:Relationship:OrganizationEmployee:CASCADE; statusTransitions:Json; defaultEmployeeRoleStatusTransitions:Json' --force`
```

File: apps/condo/domains/organization/hooks/useCreateOrganizationModalForm.tsx [hunk 18]
```diff
@@ -22,95 +27,114 @@ interface IUseCreateOrganizationModalFormProps {
...
     const CreateOrganizationModalTitle = intl.formatMessage({ id: 'pages.organizations.CreateOrganizationModalTitle' })
     const CreateOrganizationModalMsg = intl.formatMessage({ id: 'pages.organizations.CreateOrganizationMessage' })
 
     const NameMsg = intl.formatMessage({ id: 'pages.organizations.OrganizationName' })
-    const InnMessage = intl.formatMessage({ id: 'pages.organizations.Inn' })
-    const InnTooShortMsg = intl.formatMessage({ id: 'pages.organizations.inn.TooShortMessage' })
-    const mutationExtraData = {
-        country: RUSSIA_COUNTRY,
-    }
+    const InnMessage = intl.formatMessage({ id: 'pages.organizations.Tin' })
```
Comment (pahaz) [inline apps/condo/domains/organization/hooks/useCreateOrganizationModalForm.tsx:52]: tin?

File: apps/condo/domains/organization/schema/Organization.js [hunk 19]
```diff
@@ -38,6 +39,28 @@ const Organization = new GQLListSchema('Organization', {
             isRequired: true,
             kmigratorOptions: { null: false },
         },
+        // The reason for this field is to avoid adding check for resident user into global Organization read access.
+        // This field have specific use case for mobile client.
+        tin: {
+            schemaDoc: 'Organization tin',
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/Organization.js:45]: write more readable doc. Not everyone knows this abbreviation.

Comment (pahaz) [inline apps/condo/domains/organization/schema/Organization.js:45]:
```suggestion
            schemaDoc: 'Taxpayer identification number. Every country has its own identification. Examples: INN for Russia, IIN for Kazakhstan and so on',
```

File: apps/condo/domains/organization/hooks/useCreateOrganizationModalForm.tsx (hunk 18 above)
Comment (akarjakin) [reply apps/condo/domains/organization/hooks/useCreateOrganizationModalForm.tsx:52]: fixed

**Pull request merged**